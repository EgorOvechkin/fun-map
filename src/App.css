.app-container {
  display: flex;
  height: 100vh;
  padding-left: 1px;
}

.modal {
  background: #a2a29d8a;
  position: fixed;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  z-index: 999;
}

.modal .loader {
  position: absolute;
  width: 200px;
  top: 30%;
  left: calc(50% - 100px);
}

.modal .loader .message {
  width: 100%;
  text-align: center;
}

.modal .loader .spinner {
  margin: auto;
  width: 75px;
  height: 75px;
  border: 5px solid #bccf0030;
  border-radius: 50%;
  border-top-color: #638e07;
  animation: spin 1s cubic-bezier(.71,.42,.27,.6) infinite;
}

@keyframes spin {
  to { transform: rotate(360deg); }
}
