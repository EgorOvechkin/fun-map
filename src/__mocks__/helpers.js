export const generatePoints = n => Array.from({length: n}, (item, index) => ({
  title: `Point ${++index}`, coordinates: [10 * index + index, 10 * index + index]
}));
