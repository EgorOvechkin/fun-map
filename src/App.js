import React, { Component } from 'react';
import Map from './components/Map';
import RouteEditor from './components/RouteEditor';
import './App.css';

class App extends Component {
  constructor(props) {
    super(props);
    this.setMapRef = this.setMapRef.bind(this);
    this.setPointCoordinates = this.setPointCoordinates.bind(this);
    this.movePoint = this.movePoint.bind(this);
    this.addPoint = this.addPoint.bind(this);
    this.deletePoint = this.deletePoint.bind(this);
    this.state = {
      points: []
    };
  }

  render() {
    // const titles = this.state.points.map(point => point.title);

    return (
      <div className="app-container">
        <RouteEditor points={this.state.points} movePoint={this.movePoint} addPoint={this.addPoint} deletePoint={this.deletePoint}/>
        <Map points={this.state.points} onPointDrag={this.setPointCoordinates} setMapRef={this.setMapRef}/>
      </div>
    );
  }

  movePoint({oldIndex, newIndex}) {
    const array = this.state.points.slice();
    array.splice(newIndex, 0, array.splice(oldIndex, 1)[0]);

    this.setState({points: array})
  }

  setMapRef(mapRef) {
    this.map = mapRef;
  }

  setPointCoordinates(index, coordinates) {
    const points = this.state.points.slice();
    points[index].coordinates = coordinates;

    this.setState({points});
  }

  addPoint(title) {
    const point = {title, coordinates: this.map.getCenter()};

    this.setState({points: this.state.points.concat([point])})
  }

  deletePoint(index) {
    const points = this.state.points.slice();

    points.splice(index, 1);
    this.setState({points})
  }
}

export default App;
