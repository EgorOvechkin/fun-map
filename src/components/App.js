import '../App.css';
import Loader from './Loader';
import PointInput from './RouteEditor/PointInput';
import PointList from './RouteEditor/PointList';
import React, { Component } from 'react';
import RouteEditor from './RouteEditor/RouteEditor';
import YandexMap from './Map/YandexMap';

class App extends Component {
  constructor(props) {
    super(props);
    this.setMapRef = this.setMapRef.bind(this);
    this.setPointCoordinates = this.setPointCoordinates.bind(this);
    this.movePoint = this.movePoint.bind(this);
    this.addPoint = this.addPoint.bind(this);
    this.deletePoint = this.deletePoint.bind(this);
    this.state = {
      isMapLoaded: false,
      points: []
    };
  }

  render() {
    const titles = this.state.points.map(point => point.title);

    // debugger

    return (
      <div className="app-container">
        {!false && <Loader message={'Загружаем карту...'}/>}
        <RouteEditor>
          <PointInput addPoint={this.addPoint} />
          <PointList titles={titles} deletePoint={this.deletePoint} onSortEnd={this.movePoint} />
        </RouteEditor>
        <YandexMap points={this.state.points} onPointDrag={this.setPointCoordinates} setMapRef={this.setMapRef} />
      </div>
    );
  }

  movePoint({oldIndex, newIndex}) {
    const array = this.state.points.slice();
    array.splice(newIndex, 0, array.splice(oldIndex, 1)[0]);

    this.setState({points: array});
  }

  setMapRef(mapRef) {
    this.map = mapRef;
    this.setState({isMapLoaded: Boolean(mapRef)});
  }

  setPointCoordinates(index, coordinates) {
    const points = this.state.points.slice();
    points[index].coordinates = coordinates;

    this.setState({points});
  }

  addPoint(title) {
    const point = {title, coordinates: this.map.getCenter()};

    this.setState({points: this.state.points.concat([point])});
  }

  deletePoint(index) {
    const points = this.state.points.slice();

    points.splice(index, 1);
    this.setState({points});
  }
}

export default App;
