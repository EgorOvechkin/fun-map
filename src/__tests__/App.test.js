import React from 'react';
import App from '../components/App';
import {shallow} from 'enzyme';
import {generatePoints} from 'helpers';

function getAppWithPoints(length) {
  const points = generatePoints(length);
  const wrapper = shallow(<App />);

  wrapper.setState({points});

  return wrapper;
}

function getTitles(points) {
  return points.map(point => point.title).join(', ');
}

function getCoordinates(points) {
  return points.map(point => point.coordinates);
}

it('renders the app', () => {
  const wrapper = getAppWithPoints(3);
  expect(wrapper).toMatchSnapshot();
});

it('moves point', () => {
  const wrapper = getAppWithPoints(3);
  let points = wrapper.state('points');
  let titles = getTitles(points);
  let coordinates = getCoordinates(points);

  expect(points.length).toBe(3);
  expect(titles).toBe('Point 1, Point 2, Point 3');
  expect(coordinates).toMatchObject([[11,11], [22,22], [33,33]]);

  wrapper.instance().movePoint({oldIndex: 0, newIndex: 1});

  points = wrapper.state('points');
  titles = getTitles(points);
  coordinates = getCoordinates(points);
  expect(points.length).toBe(3);
  expect(titles).toBe('Point 2, Point 1, Point 3');
  expect(coordinates).toMatchObject([[22,22], [11,11], [33,33]]);
});

it('sets map ref', () => {
  const wrapper = getAppWithPoints(3);
  const ref = {offsetHeight: 100};

  wrapper.instance().setMapRef(ref);
  expect(wrapper.instance().map).toMatchObject(ref);
});

it('sets point coordinates', () => {
  const wrapper = getAppWithPoints(2);
  const newCoordinates = [33,33];
  let points = wrapper.state('points');
  let coordinates = getCoordinates(points);

  expect(coordinates).toMatchObject([[11,11], [22,22]]);

  wrapper.instance().setPointCoordinates(1, newCoordinates);

  points = wrapper.state('points');
  coordinates = getCoordinates(points);
  expect(coordinates).toMatchObject([[11,11], newCoordinates]);
});

it('adds point', () => {
  const wrapper = getAppWithPoints(1);
  const center = [55,55];
  const getCenter = jest.fn().mockReturnValue(center);
  const mapInstance = jest.fn().mockReturnValue({getCenter});

  wrapper.instance().map = mapInstance();
  wrapper.instance().addPoint('New point');
  const points = wrapper.state('points');
  const coordinates = getCoordinates(points);
  const titles = getTitles(points);

  expect(getCenter.mock.calls.length).toBe(1);
  expect(titles).toBe('Point 1, New point');
  expect(coordinates).toMatchObject([[11,11], center]);
});

it('deletes point', () => {
  const wrapper = getAppWithPoints(2);
  wrapper.instance().deletePoint(1);

  const points = wrapper.state('points');
  const coordinates = getCoordinates(points);
  const titles = getTitles(points);

  expect(titles).toBe('Point 1');
  expect(coordinates).toMatchObject([[11,11]]);
});
