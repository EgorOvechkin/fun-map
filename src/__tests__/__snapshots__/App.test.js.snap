// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the app 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <RouteEditor>
          <RouteEditor
            addPoint={[Function]}
          />
          <PointsList
            addPoint={[Function]}
            deletePoint={[Function]}
            movePoint={[Function]}
            onSortEnd={[Function]}
            titles={
              Array [
                "point_1",
                "point_2",
                "point_3",
              ]
            }
          />
        </RouteEditor>,
        <YMap
          onPointDrag={[Function]}
          points={
            Array [
              Object {
                "coordinates": Array [
                  11,
                  11,
                ],
                "title": "point_1",
              },
              Object {
                "coordinates": Array [
                  22,
                  22,
                ],
                "title": "point_2",
              },
              Object {
                "coordinates": Array [
                  33,
                  33,
                ],
                "title": "point_3",
              },
            ]
          }
          setMapRef={[Function]}
        />,
      ],
      "className": "app-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <RouteEditor
              addPoint={[Function]}
            />,
            <PointsList
              addPoint={[Function]}
              deletePoint={[Function]}
              movePoint={[Function]}
              onSortEnd={[Function]}
              titles={
                Array [
                  "point_1",
                  "point_2",
                  "point_3",
                ]
              }
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "addPoint": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "addPoint": [Function],
              "deletePoint": [Function],
              "movePoint": [Function],
              "onSortEnd": [Function],
              "titles": Array [
                "point_1",
                "point_2",
                "point_3",
              ],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onPointDrag": [Function],
          "points": Array [
            Object {
              "coordinates": Array [
                11,
                11,
              ],
              "title": "point_1",
            },
            Object {
              "coordinates": Array [
                22,
                22,
              ],
              "title": "point_2",
            },
            Object {
              "coordinates": Array [
                33,
                33,
              ],
              "title": "point_3",
            },
          ],
          "setMapRef": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <RouteEditor>
            <RouteEditor
              addPoint={[Function]}
            />
            <PointsList
              addPoint={[Function]}
              deletePoint={[Function]}
              movePoint={[Function]}
              onSortEnd={[Function]}
              titles={
                Array [
                  "point_1",
                  "point_2",
                  "point_3",
                ]
              }
            />
          </RouteEditor>,
          <YMap
            onPointDrag={[Function]}
            points={
              Array [
                Object {
                  "coordinates": Array [
                    11,
                    11,
                  ],
                  "title": "point_1",
                },
                Object {
                  "coordinates": Array [
                    22,
                    22,
                  ],
                  "title": "point_2",
                },
                Object {
                  "coordinates": Array [
                    33,
                    33,
                  ],
                  "title": "point_3",
                },
              ]
            }
            setMapRef={[Function]}
          />,
        ],
        "className": "app-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <RouteEditor
                addPoint={[Function]}
              />,
              <PointsList
                addPoint={[Function]}
                deletePoint={[Function]}
                movePoint={[Function]}
                onSortEnd={[Function]}
                titles={
                  Array [
                    "point_1",
                    "point_2",
                    "point_3",
                  ]
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "addPoint": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "addPoint": [Function],
                "deletePoint": [Function],
                "movePoint": [Function],
                "onSortEnd": [Function],
                "titles": Array [
                  "point_1",
                  "point_2",
                  "point_3",
                ],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onPointDrag": [Function],
            "points": Array [
              Object {
                "coordinates": Array [
                  11,
                  11,
                ],
                "title": "point_1",
              },
              Object {
                "coordinates": Array [
                  22,
                  22,
                ],
                "title": "point_2",
              },
              Object {
                "coordinates": Array [
                  33,
                  33,
                ],
                "title": "point_3",
              },
            ],
            "setMapRef": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
